V2/src/fc_state_machine.cpp

```diff
@@ -2,57 +2,164 @@
+{
     return state;
 }
 
-// todo - other digital pin defs
-DigitalIn start(START);
+void fc_state_machine_thread() 
+{
+    fcc_r.write(true); // Solid state relay.
 
-// todo - define which timers we need
-Timer timer;
-
-void fc_state_machine_thread() {
-    while (true) {
-        if (check_all_errors()) {
-            state = FC_FAULT;
-        }
+    while (true) 
+    {
+        if (check_all_errors()) 
+            state = FC_ALARM;
 
-        if (state == FC_TEST) {
+        if (FC_TEST == state) 
+        {
             ThisThread::sleep_for(50); // Add test state code here, if desired.
+            charge_r.write(false);
         }
 
-        if (state == FC_INIT){
-            if (start) {
-                state = FC_STARTUP;
-                // we will need lots if timers
-                // timer example:
-                timer.start();
+        if (FC_STANDBY == state) // Wait until start button is pressed.
+        { 
+            if (start) 
+            {
+                purge_v.write(false); // Perform start purge.
+                start_r.write(true);
+                motor_r.write(false);
+                charge_r.write(false);
+                cap_r.write(false);
+                supply_v.write(true);
+                Thread::wait(1000);
+                purge_v.write(true);
+                Thread::wait(200);
+                purge_v.write(false); 
+
+                supply_v.write(true); // Charge state setup.
+                start_r.write(false);
+                motor_r.write(false);
+                charge_r.write(true);
+                cap_r.write(false);
+                state = FC_CHARGE; 
+                update_leds();
             }
         }
-        else if(state == FC_STARTUP) {
-            if (timer.read() > 5.0) {
-                timer.stop();
-                //do something here
+        else if (FC_CHARGE == state) 
+        { 
+            if (get_analog_values().capvolt >= CAP_THRESHOLD) // Leave charge state.
+            {
+                supply_v.write(true); 
+                purge_v.write(false);
+                start_r.write(false);
+                motor_r.write(true);
+                charge_r.write(false);
+                cap_r.write(true);
+                state = FC_RUN;
+                update_leds();
             }
         }
-        else if(state == FC_FAULT) {
-            // Turn off all digital outputs here
+        else if (FC_RUN == state) 
+        {
+
+            if (start)
+            {
+                supply_v.write(false);
+                purge_v.write(false);
+                start_r.write(false);
+                motor_r.write(false);
+                charge_r.write(false);
+                cap_r.write(false);
+                state = FC_SHUTDOWN;
+                update_leds();
+            }
+        }
+        else if (FC_SHUTDOWN == state) 
+        {
+            // TODO Figure out so we can start/stop with start button.
```

Comment:
Doesn't what you are doing in run state allow you to shutdown with the start button?

Reply:
Need to debounce it on start or else it will immediately stop after starting

Reply:
Yes, but since we are only looking for the pin to be high (no interrupt or anything), having something in the shutdown state that resets the state machine would get executed when we don't want, I think.